### PR TITLE
Audio source sync

### PR DESCRIPTION
**Pull request by vladest**
Description:
- Logic refactoring: there was several places/methods where audio was changed. Now it will use one model with audio devices and just update UI according to the model

Test plan
- change audio source in any audio menu
- make sure other menus reflects changes
- make sure HMD audio switches to/from HMD on HMD activate/deactivate

### CODE REVIEW

**Comment by hifi-gustavo** [comment]
Build console output for commit: dde363ef3d094054e0271ab43f17c79697030a55 available at: https://gist.github.com/1ff8641358f49d1e6bfd5688abbb3b61<br/><br/>Installers available for download at: <br/>Windows: http://cachefly.highfidelity.com/HighFidelity-Beta-PR10337-dde363e.exe<br/>OS X: http://cachefly.highfidelity.com/HighFidelity-Beta-PR10337-dde363e.dmg

**Comment by AndrewMeadows** [comment]
rebuild this please

**Comment by hifi-gustavo** [comment]
Build console output for commit: dde363ef3d094054e0271ab43f17c79697030a55 available at: https://gist.github.com/1ff8641358f49d1e6bfd5688abbb3b61<br/><br/>Installers available for download at: <br/>Windows: http://cachefly.highfidelity.com/HighFidelity-Beta-PR10337-dde363e.exe<br/>OS X: http://cachefly.highfidelity.com/HighFidelity-Beta-PR10337-dde363e.dmg

**Comment by sethalves** [comment]
test this please

**Comment by hifi-gustavo** [comment]
Build console output for commit: dde363ef3d094054e0271ab43f17c79697030a55 available at: https://gist.github.com/1ff8641358f49d1e6bfd5688abbb3b61<br/><br/>Installers available for download at: <br/>Windows: http://cachefly.highfidelity.com/HighFidelity-Beta-PR10337-dde363e.exe<br/>OS X: http://cachefly.highfidelity.com/HighFidelity-Beta-PR10337-dde363e.dmg

**Comment by hifi-gustavo** [comment]
Build console output for commit: e789b69a8e8d53681adc7116b5cffe97c500d006 available at: https://gist.github.com/1ff8641358f49d1e6bfd5688abbb3b61<br/><br/>Installers available for download at: <br/>Windows: http://cachefly.highfidelity.com/HighFidelity-Beta-PR10337-e789b69.exe<br/>OS X: http://cachefly.highfidelity.com/HighFidelity-Beta-PR10337-e789b69.dmg

**Comment by cozza13** [comment]
ok to test

**Comment by hifi-gustavo** [comment]
Build console output for commit: e789b69a8e8d53681adc7116b5cffe97c500d006 available at: https://gist.github.com/1ff8641358f49d1e6bfd5688abbb3b61<br/><br/>Installers available for download at: <br/>Windows: http://cachefly.highfidelity.com/HighFidelity-Beta-PR10337-e789b69.exe<br/>OS X: http://cachefly.highfidelity.com/HighFidelity-Beta-PR10337-e789b69.dmg

File: interface/src/scripting/AudioDeviceScriptingInterface.cpp [hunk 1]
```diff
@@ -51,13 +57,27 @@ bool AudioDeviceScriptingInterface::setInputDevice(const QString& deviceName) {
 
 bool AudioDeviceScriptingInterface::setOutputDevice(const QString& deviceName) {
     bool result;
+    QTime a; a.start();
```
Comment (sethalves) [inline interface/src/scripting/AudioDeviceScriptingInterface.cpp]: variable names should be descriptive

File: interface/src/scripting/AudioDeviceScriptingInterface.cpp [hunk 2]
```diff
@@ -51,13 +57,27 @@ bool AudioDeviceScriptingInterface::setInputDevice(const QString& deviceName) {
...
     QMetaObject::invokeMethod(DependencyManager::get<AudioClient>().data(), "switchOutputToAudioDevice",
                               Qt::BlockingQueuedConnection,
                               Q_RETURN_ARG(bool, result),
                               Q_ARG(const QString&, deviceName));
+    qDebug() << "switching to" << deviceName << "elapsed" << a.elapsed();
```
Comment (sethalves) [inline interface/src/scripting/AudioDeviceScriptingInterface.cpp]: maybe should be a qCInfo() ?

File: interface/src/scripting/AudioDeviceScriptingInterface.cpp [hunk 3]
```diff
@@ -116,3 +136,94 @@ void AudioDeviceScriptingInterface::setMuted(bool muted)
...
+    currentDeviceUpdate(getOutputDevice(), QAudio::AudioOutput);
+    emit currentOutputDeviceChanged();
+}
+
+void AudioDeviceScriptingInterface::currentDeviceUpdate(const QString &name, QAudio::Mode mode)
+{
+    QVector<int> role;
+    role.append(SelectedRole);
+
+    qDebug() << "device update" << name << mode;
```
Comment (sethalves) [inline interface/src/scripting/AudioDeviceScriptingInterface.cpp]: qCInfo() or remove?

File: interface/src/scripting/AudioDeviceScriptingInterface.cpp [hunk 4]
```diff
@@ -116,3 +136,94 @@ void AudioDeviceScriptingInterface::setMuted(bool muted)
...
+        if (di.selected && di.name != name ) {
+            di.selected = false;
+            _devices[i] = di;
+            emit dataChanged(index(i, 0), index(i, 0), role);
+        }
+        if (di.name == name) {
+            di.selected = true;
+            _devices[i] = di;
+            emit dataChanged(index(i, 0), index(i, 0), role);
+            qDebug() << "device updated" << di.name << i;
```
Comment (sethalves) [inline interface/src/scripting/AudioDeviceScriptingInterface.cpp]: qCInfo() ?

**Comment by hifi-gustavo** [comment]
Build console output for commit: d522864141bf3e53a8dbf6fa15899d5645a5fd43 available at: https://gist.github.com/1ff8641358f49d1e6bfd5688abbb3b61<br/><br/>Installers available for download at: <br/>Windows: http://cachefly.highfidelity.com/HighFidelity-Beta-PR10337-d522864.exe<br/>OS X: http://cachefly.highfidelity.com/HighFidelity-Beta-PR10337-d522864.dmg

**Comment by cozza13** [comment]
- the issue that makes this unable to be merged. 
 - Open the tablet and change the audio
 - Open the menu from the menu bar and see the audio changed (correct!)
 - Now change the audio from the menu. 
 - Reopen the tablet and go to audio and try and change the audio (You are now unable to change the audio)

**Comment by vladest** [comment]
Try again. If it still fails, please send me a logs

**Comment by hifi-gustavo** [comment]
Build console output for commit: fa9a4053cabc9847218625bc952ba4367e39e719 available at: https://gist.github.com/1ff8641358f49d1e6bfd5688abbb3b61<br/><br/>Installers available for download at: <br/>Windows: http://cachefly.highfidelity.com/HighFidelity-Beta-PR10337-fa9a405.exe<br/>OS X: http://cachefly.highfidelity.com/HighFidelity-Beta-PR10337-fa9a405.dmg

**Comment by zzmp** [comment]
Can you update the title or description to give an overview of what this is fixing or changing?

**Comment by vladest** [comment]
@zzmp logic refactoring. there was several places/methods where audio was changed. Now it will use one model with audio devices and just update UI according to the model

**Comment by zzmp** [comment]
@vladest can you put that in the description?

**Comment by hifi-gustavo** [comment]
Build console output for commit: 4290eb21eaf3459f615c6149a7e35cb86e04f371 available at: https://gist.github.com/1ff8641358f49d1e6bfd5688abbb3b61<br/><br/>Installers available for download at: <br/>Windows: http://cachefly.highfidelity.com/HighFidelity-Beta-PR10337-4290eb2.exe<br/>OS X: http://cachefly.highfidelity.com/HighFidelity-Beta-PR10337-4290eb2.dmg

**Comment by cozza13** [comment]
I just tested this and it fixes the issue

File: interface/resources/qml/hifi/Audio.qml [hunk 5]
```diff
@@ -191,31 +192,33 @@ Rectangle {
...
+            delegate: Item {
                 width: parent.width
-                checkbox.checked: (modelData === AudioDevice.getOutputDevice())
-                text.text: modelData
-                onCheckBoxClicked: {
-                    if (checked) {
-                        AudioDevice.setOutputDevice(modelData)
-                        refreshTimerOutput.start()
+                visible: devicemode === 1
+                height:  visible ? 36 : 0
```
Comment (sethalves) [inline interface/resources/qml/hifi/Audio.qml]: stray space

File: interface/src/scripting/AudioDeviceScriptingInterface.cpp [hunk 6]
```diff
@@ -11,33 +11,47 @@
...
-        this, &AudioDeviceScriptingInterface::deviceChanged);
+        this, &AudioDeviceScriptingInterface::onDeviceChanged, Qt::QueuedConnection);
+    connect(DependencyManager::get<AudioClient>().data(), &AudioClient::currentInputDeviceChanged,
+        this, &AudioDeviceScriptingInterface::onCurrentInputDeviceChanged, Qt::QueuedConnection);
+    connect(DependencyManager::get<AudioClient>().data(), &AudioClient::currentOutputDeviceChanged,
+        this, &AudioDeviceScriptingInterface::onCurrentOutputDeviceChanged, Qt::QueuedConnection);
+    //fill up model
+    onDeviceChanged();
+    //set up previously saved device
+    SettingsScriptingInterface *settings = SettingsScriptingInterface::getInstance();
```
Comment (sethalves) [inline interface/src/scripting/AudioDeviceScriptingInterface.cpp]: \* should be next to the type, like `SettingsScriptingInterface* settings`

File: interface/src/scripting/AudioDeviceScriptingInterface.cpp [hunk 7]
```diff
@@ -116,3 +167,104 @@ void AudioDeviceScriptingInterface::setMuted(bool muted)
...
+
+    endResetModel();
+    emit deviceChanged();
+}
+
+void AudioDeviceScriptingInterface::onCurrentInputDeviceChanged(const QString &name)
+{
+    currentDeviceUpdate(name, QAudio::AudioInput);
+    //we got a signal that device changed. Save it now
+    SettingsScriptingInterface *settings = SettingsScriptingInterface::getInstance();
```
Comment (sethalves) [inline interface/src/scripting/AudioDeviceScriptingInterface.cpp]: `SettingsScriptingInterface* settings`

File: interface/src/scripting/AudioDeviceScriptingInterface.cpp [hunk 8]
```diff
@@ -116,3 +167,104 @@ void AudioDeviceScriptingInterface::setMuted(bool muted)
...
+    settings->setValue("audio_input_device", name);
+    emit currentInputDeviceChanged(name);
+}
+
+void AudioDeviceScriptingInterface::onCurrentOutputDeviceChanged(const QString &name)
```
Comment (sethalves) [inline interface/src/scripting/AudioDeviceScriptingInterface.cpp]: `const QString& name`

File: interface/src/scripting/AudioDeviceScriptingInterface.cpp [hunk 9]
```diff
@@ -58,6 +72,43 @@ bool AudioDeviceScriptingInterface::setOutputDevice(const QString& deviceName) {
     return result;
 }
 
+bool AudioDeviceScriptingInterface::setDeviceFromMenu(const QString &deviceMenuName) {
```
Comment (sethalves) [inline interface/src/scripting/AudioDeviceScriptingInterface.cpp]: `const QString& deviceMenuName`

File: interface/src/scripting/AudioDeviceScriptingInterface.cpp [hunk 10]
```diff
@@ -116,3 +167,104 @@ void AudioDeviceScriptingInterface::setMuted(bool muted)
...
+        _devices.append(di);
+        _inputAudioDevices.append(name);
+    }
+    emit inputAudioDevicesChanged(_inputAudioDevices);
+
+    endResetModel();
+    emit deviceChanged();
+}
+
+void AudioDeviceScriptingInterface::onCurrentInputDeviceChanged(const QString &name)
```
Comment (sethalves) [inline interface/src/scripting/AudioDeviceScriptingInterface.cpp]: `const QString& name`

File: interface/src/scripting/AudioDeviceScriptingInterface.cpp [hunk 11]
```diff
@@ -116,3 +167,104 @@ void AudioDeviceScriptingInterface::setMuted(bool muted)
...
+void AudioDeviceScriptingInterface::onCurrentOutputDeviceChanged(const QString &name)
+{
+    currentDeviceUpdate(name, QAudio::AudioOutput);
+    //we got a signal that device changed. Save it now
+    SettingsScriptingInterface *settings = SettingsScriptingInterface::getInstance();
+    settings->setValue("audio_output_device", name);
+    emit currentOutputDeviceChanged(name);
+}
+
+void AudioDeviceScriptingInterface::currentDeviceUpdate(const QString &name, QAudio::Mode mode)
```
Comment (sethalves) [inline interface/src/scripting/AudioDeviceScriptingInterface.cpp]: const QString& name

File: interface/src/scripting/AudioDeviceScriptingInterface.h [hunk 12]
```diff
@@ -32,9 +40,26 @@ class AudioDeviceScriptingInterface : public QObject {
...
+    enum Roles {
+        DisplayNameRole = Qt::UserRole,
+        SelectedRole,
+        AudioModeRole
+    };
+
+private slots:
+    void onDeviceChanged();
+    void onCurrentInputDeviceChanged(const QString &name);
+    void onCurrentOutputDeviceChanged(const QString &name);
```
Comment (sethalves) [inline interface/src/scripting/AudioDeviceScriptingInterface.h]: const QString& name

File: interface/src/scripting/AudioDeviceScriptingInterface.h [hunk 13]
```diff
@@ -32,9 +40,26 @@ class AudioDeviceScriptingInterface : public QObject {
...
+    void currentDeviceUpdate(const QString &name, QAudio::Mode mode);
```
Comment (sethalves) [inline interface/src/scripting/AudioDeviceScriptingInterface.h]: and here

File: interface/src/scripting/AudioDeviceScriptingInterface.h [hunk 14]
```diff
@@ -55,15 +80,28 @@ public slots:
     
     void setMuted(bool muted);
 
+    void setInputDeviceAsync(const QString &deviceName);
+    void setOutputDeviceAsync(const QString &deviceName);
```
Comment (sethalves) [inline interface/src/scripting/AudioDeviceScriptingInterface.h]: QString&

File: interface/src/scripting/AudioDeviceScriptingInterface.h [hunk 15]
```diff
@@ -55,15 +80,28 @@ public slots:
...
 private:
     AudioDeviceScriptingInterface();
 
 signals:
     void muteToggled();
     void deviceChanged();
+    void currentInputDeviceChanged(const QString &name);
+    void currentOutputDeviceChanged(const QString &name);
```
Comment (sethalves) [inline interface/src/scripting/AudioDeviceScriptingInterface.h]: QString&

File: libraries/audio-client/src/AudioClient.h [hunk 16]
```diff
@@ -222,6 +222,9 @@ public slots:
 
     void muteEnvironmentRequested(glm::vec3 position, float radius);
 
+    void currentOutputDeviceChanged(const QString &name);
+    void currentInputDeviceChanged(const QString &name);
```
Comment (sethalves) [inline libraries/audio-client/src/AudioClient.h]: QString&

File: interface/src/scripting/AudioDeviceScriptingInterface.cpp [hunk 17]
```diff
@@ -116,3 +167,104 @@ void AudioDeviceScriptingInterface::setMuted(bool muted)
...
+    SettingsScriptingInterface *settings = SettingsScriptingInterface::getInstance();
+    settings->setValue("audio_input_device", name);
+    emit currentInputDeviceChanged(name);
+}
+
+void AudioDeviceScriptingInterface::onCurrentOutputDeviceChanged(const QString &name)
+{
+    currentDeviceUpdate(name, QAudio::AudioOutput);
+    //we got a signal that device changed. Save it now
+    SettingsScriptingInterface *settings = SettingsScriptingInterface::getInstance();
```
Comment (sethalves) [inline interface/src/scripting/AudioDeviceScriptingInterface.cpp]: `SettingsScriptingInterface* settings`

File: interface/src/scripting/AudioDeviceScriptingInterface.cpp [hunk 18]
```diff
@@ -116,3 +167,104 @@ void AudioDeviceScriptingInterface::setMuted(bool muted)
...
+}
+
+void AudioDeviceScriptingInterface::currentDeviceUpdate(const QString &name, QAudio::Mode mode)
+{
+    QVector<int> role;
+    role.append(SelectedRole);
+
+    for (int i = 0; i < _devices.size(); i++) {
+        ScriptingAudioDeviceInfo di = _devices.at(i);
+        if (di.mode != mode)
```
Comment (sethalves) [inline interface/src/scripting/AudioDeviceScriptingInterface.cpp]: coding standard requires { and } even for 1-liners

File: libraries/script-engine/src/AudioScriptingInterface.h [hunk 19]
```diff
@@ -24,6 +24,7 @@ class AudioScriptingInterface : public QObject, public Dependency {
     SINGLETON_DEPENDENCY
 
 public:
+        virtual ~AudioScriptingInterface() {}
```
Comment (sethalves) [inline libraries/script-engine/src/AudioScriptingInterface.h]: strange indentation?

File: scripts/system/selectAudioDevice.js [hunk 20]
```diff
@@ -255,21 +187,24 @@ function checkHMDAudio() {
...
+    AudioDevice.currentOutputDeviceChanged.connect(onCurrentDeviceChanged);
     HMD.displayModeChanged.connect(checkHMDAudio);
     Menu.menuItemEvent.connect(onMenuEvent);
     debug("Setting up Audio I/O menu for the first time...");
     setupAudioMenus();
-    checkDeviceMismatch();
     debug("Checking HMD audio status...")
     checkHMDAudio();
-}, 3000);
+}, 300);
```
Comment (sethalves) [inline scripts/system/selectAudioDevice.js]: magic number.  store 300 in a descriptively named variable.

**Pull request merged**